### DISCUSSION TITLE
Constrain nodes

### DISCUSSION THREAD

**Discussion by dominikkau** [ImperialCollegeLondon/sharpy#181] [Q&A]
Hi,
is it possible to constrain individual degrees of freedoms of specific nodes? If not, is this a feature that xbeam does not provide or is only the interface in sharpy missing?
As far as I can see it is only possible to constrain all degrees of freedom at once by setting boundary_conditions = 1 for the reference node. I would like to allow rotational degrees of freedom while restricting the translational ones. At the moment I am trying to achieve this by adding beams with very high stiffness - except e.g. torsional stiffness - from the reference node to the node that I want to restrict. However I am not convinced that this method will produce correct results especially for modal analysis.
Thanks in advance!

**Comment by ACea15** [comment]
Hi,
SHARPy structural solver (xbeam) is general but not a fully fledged FE solver, you will not be able to constrain individual DoF in arbitrary nodes, however, you can achieve what you describe using the multibody capability of the code! The wind turbines examples do just that, clamping the tower to the rotor but allowing rotational DoF between them. 
I don't see any problem with the solution you propose (effectively a CBUSH in Nastran but with a bit of noise from the numerics that are not removed). It might be quicker to implement. 
Finally, it is better if you describe the actual problem you are trying to solve, since there might already be a solver that does it for you. Bear in mind that BC=1 only defines the A frame node that is then clamped, or not, depending on the problem at hand defined through the software workflow.

**Reply by dominikkau** [reply]
Hi,
thanks for your answer! We want to simulate a wing and wing box as a single beam. I was talking about the mount of the wingbox to the fuselage at the wingroot which we assume to be at a fixed position because of the stiff fuselage but we want to allow a rotation of the wing in the root. This difference obviously has a large influence on the eigenmodes of the system which are our primary interest at the moment.
With regard to the multibody capabilities: I was under the impression that most of the solvers don't support this functionality.

**Reply by ArturoMS13** [reply]
Hi Domino0o,
The multibody features are available for nonlinear dynamic simulations. You can see examples on how to constrain individual node linear velocity in "tests/coupled/multibody". This is not a feature of xbeam, but a multibody implementation on top of it.

I agree with ACea15 that the solution with increased stiffness elements might be simpler. Moreover, the use of the modal solutions with SHARPy on multibody systems is not easy.

**Comment by ACea15** [comment]
I dont UNDERSTAND what you want to model Domino!! Your description is of a no fixed-wing aircraft (fuselage and wing not being fully attached). In that case you would need actuation for sure, or would you let the wings to spin freely?? In military aircraft they do have the capability to actuate on the wing sweep at the root to control compressible effects; and birds flap to generate extra lift...  
If on the contrary you just want to set the wing at a particular initial AoA, the solution is not having decoupled DoF but using the SHARPy variables twist or frame_of_reference_delta.

**Reply by dominikkau** [reply]
We want to model a flexible connection between fuselage and wing in the bending direction of the wing. So I am talking about a rotation around the x-axis of the A frame in sharpy. We are modelling a fixed wing aircraft but we want to take into account that the wing root is not a completly rigid joint.

To further clarify our intention, I attached an image viewing the airplane in a plane parallel to the x-axis. The classical approach is on top where the wing is connected to the fuselage in a fully rigid manner. This can also be simulated with a half span model and all eigenmodes would be symmetrical. We want to model the wing and the wing box as one single beam that has a fixed position in the wing root. This then allows for determination of eigenmodes that are not symmetric (such as the blue one I drew)

![great_drawing_of_airplane](https://user-images.githubusercontent.com/51756125/144908845-7ff75f2d-eb9c-4acb-8ab6-0653c5a1564b.jpg)

**Reply by ACea15** [reply]
Ohh I see now, then you don't want to decouple DoF. Have your reference A frame node at the centre of the fuselage, beam components from there to the sides  (your black dots), and from there the wing beam components.  The more nodes, or elements, you put in between the black dots to model the wing-box, the more resolution you will have there, hence more harmonics in the blue line you have plotted. Also the less stiff those elements are, the smaller the frequencies associated with that part of the model and possibly couplings with the wing dynamics may appear.
SHARPy modal solution should be able to capture this. 
Regards

**Reply by dominikkau** [reply]
Hi ACea15, thanks for you answer!
Unfortunately I am not quite sure if i understand it correctly.
With the beam from the reference A frame node to the black dots you don't mean the wing box but an "abstracted fuselage", right?
Would you use multibodies at the intesection point or do you still think using soft beams is still the better option?
Regards

**Reply by ngoiz** [reply]
Hi @DomIno0o,

If I may chip in to what @ACea15 has said, in SHARPy, we sometimes refer to the A frame as a clamped node (boundary conditions = 1 at that node). However, in reality this is just a reference node and, of course, if we want to simulate any flying vehicle that node is anything but clamped. Therefore, if you run a nonlinear time domain simulation, the dynamics of the wing will be those of a single beam (wingbox) from wingtip to wingtip, with symmetric and antisymmetric modes. 

If you want to visualise these modes, you can run the Modal solver with the setting to simulate `rigid_body_motion = True`. That way you will have the first 10 modes with no shape (rigid body velocities, angular velocities and quaternion) and then the flexible modes, which will be like the blue ones you have drawn. To complete the answer, if you set `rigid_body_motion = False` in the modal solver, then it will be assumed that the wing is clamped and then you will get modes for the right and left wings, which you could add in post-processing if needed.

If you want to model different properties on the wingbox before and after the fuselage, then is as @ACea15 has said a matter of adding elements with different properties and the more you add the more refined/gradual changes you will be able to make.

Hope this helps,
Norberto

**Reply by dominikkau** [reply]
Hi @ngoiz,

after reading your message I realised that maybe I misunderstood @ACea15s message. Still, I can't see how I can achieve the results I want with the method you are describing. The black dots in the upper image are meant to be clamped except for rotation around the x-axis parallel to the fuselage (obviously the points are also moving with a constant cruise speed but that is a solver setting as far as I understand). Maybe I am wrong but my understanding is, that due to the coupling with aerodynamics our model is not the same as a free beam that you obtain when setting just the middle node as the reference node and allowing rigid body motion.
I'm sorry that the discussion is dragging on like this...
